Added the ability to force the use of an index

snake14 opened the pull request:
### Description:

Added the ability to force the use of an index when aggregating logs. This is to try and improve efficiency when the query engine doesn't always automatically select the correct index.

Internal ticket: PG-3720

### Review

* [ ] [Functional review done](https://developer.matomo.org/guides/pull-request-reviews#functional-review-done)
* [ ] [Potential edge cases thought about](https://developer.matomo.org/guides/pull-request-reviews#potential-edge-cases-thought-about) (behavior of the code with strange input, with strange internal state or possible interactions with other Matomo subsystems)
* [ ] [Usability review done](https://developer.matomo.org/guides/pull-request-reviews#usability-review-done) (is anything maybe unclear or think about anything that would cause people to reach out to support)
* [ ] [Security review done](https://developer.matomo.org/guides/security-in-piwik#checklist)
* [ ] [Wording review done](https://developer.matomo.org/guides/pull-request-reviews#translations-wording-review-done)
* [ ] [Code review done](https://developer.matomo.org/guides/pull-request-reviews#code-review-done)
* [ ] [Tests were added if useful/possible](https://developer.matomo.org/guides/pull-request-reviews#tests-were-added-if-usefulpossible)
* [ ] [Reviewed for breaking changes](https://developer.matomo.org/guides/pull-request-reviews#reviewed-for-breaking-changes)
* [ ] [Developer changelog updated if needed](https://developer.matomo.org/guides/pull-request-reviews#developer-changelog-updated-if-needed)
* [ ] [Documentation added if needed](https://developer.matomo.org/guides/pull-request-reviews#documentation-added-if-needed)
* [ ] Existing documentation updated if needed

snake14 committed:
Added the ability to force the use of an index

snake14 committed:
Improving test a little more

michalkleiner commented on core/DataAccess/LogAggregator.php:
```diff
@@ -1176,6 +1176,8 @@ protected function getActionsMetricFields()
      * @param bool $rankingQueryGenerate if `true`, generates a SQL query / bind array pair and returns it. If false, the
      *                                   ranking query SQL will be immediately executed and the results returned.
+     * @param bool $forceSiteDateIndex Forces the resulting query to use the index_idsite_datetime index. For some
+     * reason, the engine doesn't always use that index automatically. This allows use to make sure that it uses it.
```
```suggestion
     * reason, the engine doesn't always use that index automatically. This allows us to make sure that it uses it.
```

michalkleiner commented on core/DataAccess/LogQueryBuilder/JoinTables.php:
```diff
@@ -155,6 +155,11 @@ public function sort()
         $firstTable = array_shift($tables);
         $sorted = [$firstTable];
 
+        // Added the ability to specify which index to use, so the first table might be an array
```
```suggestion
        // With the ability to specify which index to use, the $firstTable var may be an array
```

snake14 committed:
Make suggested comment correction

Co-authored-by: Michal Kleiner <michal@innocraft.com>

snake14 committed:
Make suggested comment correction

Co-authored-by: Michal Kleiner <michal@innocraft.com>

snake14 commented:
Thank you for the input @michalkleiner . I have committed your suggested changes.

sgiehl commented:
@snake14 Did you check the resulting queries on a bigger database using different segment types and also on mysql and mariadb? We need to ensure that forcing an index doesn't slow down anything. Also mysql and mariadb might behave differently in terms of query optimisation.
Maybe there could also be other ways to get mysql to use the index, e.g. by changing order of where conditions or similar.
In the end we need some reliable numbers to be able to merge that.

snake14 commented:
> @snake14 Did you check the resulting queries on a bigger database using different segment types and also on mysql and mariadb? We need to ensure that forcing an index doesn't slow down anything. Also mysql and mariadb might behave differently in terms of query optimisation. Maybe there could also be other ways to get mysql to use the index, e.g. by changing order of where conditions or similar. In the end we need some reliable numbers to be able to merge that.

@sgiehl I haven't tested these specific queries with a large database, but a [similar change has been live for the Funnels plugin](https://github.com/innocraft/plugin-Funnels/pull/187) for a few months. This PR is based on those findings. I suppose I could put this feature behind a config, have it deployed to Cloud using a patch, and profile a customer's archiving process. What do you think @patrickli ?

patrickli commented:
yeah sure why not

snake14 committed:
Adding new config to control use of new feature

snake14 committed:
Fixing test after recent change

snake14 committed:
Fixing test to avoid affecting other tests

snake14 committed:
Fixing PHPCS error

snake14 committed:
Applying some PR review feedback from Cloud team

snake14 commented:
> This issue is in "needs review" but there has been no activity for 7 days. ping @matomo-org/core-reviewers

These changes have been put into a [patch for Cloud](https://github.com/innocraft/matomo-cloud/pull/2499). It's been merged and should be deployed this week. We'll then activate the changes for a customer, see if we can profile their archiving process, and compare to previous profiling results.

snake14 committed:
Fixing edge case

snake14 committed:
Adding test case around regression found in patch on Cloud

snake14 committed:
Making slight test case improvement

snake14 commented:
@sgiehl The change didn't appear to make a difference for a couple plugins, but it appeared to make an improvement for CustomVariables. Please [see this comment](https://innocraft.atlassian.net/browse/PG-3858?focusedCommentId=198703) on our internal ticket. I believe that profiling was done using the site of one of our larger Cloud customers. Is that good enough for merging or is there something else that you would like to see tested?

snake14 commented:
@sgiehl Just trying to check in again if we'd be able to merge this PR? Cloud testing went well. When profiling a customer's site with the index change enabled/disabled, two plugins had no noticeable difference (likely already using index) while one plugin had a reasonable improvement in performance. As no harm to performance was noticed and there was some improvement, I think that we're probably safe to merge this, especially since it's only used on a case-by-case basis and not all queries.

sgiehl commented:
@snake14 should we remove the config option to enable/disable that again? If it was well tested and it works as expected, I don't think the config option will be of use anymore.

snake14 commented:
> @snake14 should we remove the config option to enable/disable that again? If it was well tested and it works as expected, I don't think the config option will be of use anymore.

@sgiehl I'd kind of like to keep it if possible. I've currently got several plugin PRs checking for `method_exists(SettingsPiwik::class, 'isForceSiteDateIndexEnabled')` so that we don't have to increase the minimum required Matomo version for the plugin.

sgiehl commented:
@snake14 I still don't see any value in keeping that config check code. If there is no reason for the need to disable it, then there is no need to keep the code.
And in terms of plugins I don't see the problem. Didn't you 'only' introduce an additional parameter to a method? So if you always pass that new parameter to the method, the parameter would be simply ignored from Matomo versions where it was not yet defined, right?

snake14 commented:
@sgiehl That makes sense. I don't see any harm in leaving it and defaulting it to on in the future, though.

I don't really like the idea of making method calls that don't match the method signature; maybe it's just from my years working with a more strict language like Java. However, I have other plugin PRs that are using `LogAggregator->generateQuery()` directly and need to be able to know whether or not to pass the from table as an array or not. Otherwise, it would cause type exceptions. I was trying to avoid making more Core changes as this PR has taken quite a bit of time already.

sgiehl commented:
In your plugins you can also simply check for the Matomo version this is included in

snake14 committed:
Removed config for forcing index

snake14 committed:
Removed unused import

snake14 commented:
> In your plugins you can also simply check for the Matomo version this is included in

@sgiehl Good point. I forgot that I had Matomo version checks in my PRs before I switched to the config. I have removed the config.

sgiehl commented on tests/PHPUnit/Unit/DataAccess/LogAggregatorTest.php:
```diff
@@ -0,0 +1,101 @@
+        $dbMock->expects($this->once())->method('query')->with($this->stringContains($expectedSql), $this->equalTo([]));
+
+        Config::getInstance()->General['enable_force_site_date_index'] = 1;
+
```
This should no longer be needed.

```suggestion
```

sgiehl committed:
Remove no longer needed code

sgiehl committed:
remove unneeded use statement

sgiehl reviewed (approved):
This looks fine to me now.
